src/aiida/manage/configuration/settings.py

```diff
@@ -25,70 +26,98 @@
 DEFAULT_DAEMON_LOG_DIR_NAME = 'log'
 DEFAULT_ACCESS_CONTROL_DIR_NAME = 'access'
 
-# Assign defaults which may be overriden in set_configuration_directory() below
-AIIDA_CONFIG_FOLDER: pathlib.Path = pathlib.Path(DEFAULT_AIIDA_PATH).expanduser() / DEFAULT_CONFIG_DIR_NAME
-DAEMON_DIR: pathlib.Path = AIIDA_CONFIG_FOLDER / DEFAULT_DAEMON_DIR_NAME
-DAEMON_LOG_DIR: pathlib.Path = DAEMON_DIR / DEFAULT_DAEMON_LOG_DIR_NAME
-ACCESS_CONTROL_DIR: pathlib.Path = AIIDA_CONFIG_FOLDER / DEFAULT_ACCESS_CONTROL_DIR_NAME
+__all__ = ('AiiDAConfigPathResolver', 'AiiDAConfigDir')
 
 
-def create_instance_directories() -> None:
+@final
+class AiiDAConfigDir:
+    """Singleton for setting and getting the path to configuration directory."""
```

Comment:
```suggestion
    """Global access class that hold the attribute for the absolute path of AiiDA configuration folder. The class method ``set`` and ``get`` are provided to mutate and access the location of configuration folder."""
```

Look back and the implementation, I think it is not the actually singleton pattern, which only instantiate the object once and only once for entire program. 
The old change I put was having a `_glb_aiida_config_folder` variable and it seems fit more with the global variable of the module and the module in python is a singleton so the global variable will therefore shared over the entire program. 
In the current implementation, it is `AiiDAconfigDir._glb_aiida_config_folder` the class attribute plays this global variable role. 

I think the new docstring above makes more sense for the class.

Reply:
It is indeed definitely not a singleton. I think I left a comment before asking why you went with a class and a class attribute to store the global config directory, but I cannot find it nor a response. I still don't really see why this requires a class? Wouldn't it be perfectly fine to simply do:
```python
_AIIDA_CONFIG_FOLDER: pathlib.Path | None = None

def get():
    global _AIIDA_CONFIG_FOLDER
    return _AIIDA_CONFIG_FOLDER

def set(aiida_config_folder):
    global _AIIDA_CONFIG_FOLDER
    _AIIDA_CONFIG_FOLDER = aiida_config_folder
    _create_instance_directories(aiida_config_folder)
```

Reply:
I think I implemented in this way and then agree with @danielhollas it would be better to move as class attribute to the class. 
Logically it is the same as current implementation, just the difference of how codes are organized and how the set/get functions are called. 

With class and attributes, the class server as the global variable with the module, and get/set are called by:

```python
from aiida.manage.configuration.settings import AiiDAconfigDir

AiiDAconfigDir.set("/tmp/.aiida")
config_folder = AiiDAconfigDir.get()
```

with global variable of module, get/set are called by:

```python
from aiida.manage.configuation import settings

settings.set_configuration_folder("/tmp/.aiida")
config_folder = settings.get_configuration_folder()
```

I personally prefer the `AiiDAConfigDir` way which seems more compact.

Reply:
Sure, in the end it is a matter of taste. I just wanted to ask because I don't like this general attitude that "everything must be a class". There are plenty of situations where creating an entire class for a simple function is not necessary.

Reply:
There is another advantage of using `global` keyword which is I have to put the `# PLW0608` to silent the mypy. It makes it easier to find where we use the global variables. Usually those are places that may not thread-safe. 

@danielhollas want to hear what you think, I think after separate the `AiiDAConfigPathResover` out, it seems better to not using `AiiDAConifgDir` to encapsulate it as @sphuber suggested?